### PR TITLE
Implement Snipe Shot, and Propeller Tail/Stalwart

### PR DESCRIPTION
**Pull request by f-raZ0R**
This pull request adds the functionality of ignoring redirects to the move Snipe Shot, and the abilities Propeller Tail and Stalwart. Currently, the only things that can redirect a move, are the abilities Storm Drain and Lightning Rod. If more ways to do so get implemented in the future, these moves and abilities may need to be tweaked further.

### CODE REVIEW

**Review by bennybroseph — CHANGES_REQUESTED** [review]

File: src/phases.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ import { Weather, WeatherType, getRandomWeatherType, getTerrainBlockMessage, get
...
+import { CheckTrappedAbAttr, IgnoreOpponentStatChangesAbAttr, PostAttackAbAttr, PostBattleAbAttr, PostDefendAbAttr, PostSummonAbAttr, PostTurnAbAttr, PostWeatherLapseAbAttr, PreSwitchOutAbAttr, PreWeatherDamageAbAttr, ProtectStatAbAttr, RedirectMoveAbAttr, BlockRedirectAbAttr, RunSuccessAbAttr, StatChangeMultiplierAbAttr, SuppressWeatherEffectAbAttr, SyncEncounterNatureAbAttr, applyAbAttrs, applyCheckTrappedAbAttrs, applyPostAttackAbAttrs, applyPostBattleAbAttrs, applyPostDefendAbAttrs, applyPostSummonAbAttrs, applyPostTurnAbAttrs, applyPostWeatherLapseAbAttrs, applyPreStatChangeAbAttrs, applyPreSwitchOutAbAttrs, applyPreWeatherEffectAbAttrs, BattleStatMultiplierAbAttr, applyBattleStatMultiplierAbAttrs, IncrementMovePriorityAbAttr, applyPostVictoryAbAttrs, PostVictoryAbAttr, applyPostBattleInitAbAttrs, PostBattleInitAbAttr, BlockNonDirectDamageAbAttr as BlockNonDirectDamageAbAttr, applyPostKnockOutAbAttrs, PostKnockOutAbAttr, PostBiomeChangeAbAttr, applyPostFaintAbAttrs, PostFaintAbAttr, IncreasePpAbAttr, PostStatChangeAbAttr, applyPostStatChangeAbAttrs, AlwaysHitAbAttr, PreventBerryUseAbAttr, StatChangeCopyAbAttr } from "./data/ability";
```
[the remark this answers is not in the document]
Comment (f-raZ0R) [reply src/phases.ts:33]: BlockRedirectAbAttr was added to the attributes imported

File: src/phases.ts [hunk 2]
```diff
@@ -2210,8 +2210,11 @@ export class MovePhase extends BattlePhase {
     const moveTarget = this.targets.length === 1
       ? new Utils.IntegerHolder(this.targets[0])
       : null;
-    if (moveTarget) {
-      this.scene.getField(true).filter(p => p !== this.pokemon).forEach(p => applyAbAttrs(RedirectMoveAbAttr, p, null, this.move.moveId, moveTarget));
+      if (moveTarget) {
+          if (!(this.pokemon.hasAbilityWithAttr(BlockRedirectAbAttr) || this.move.getMove().getAttrs(BypassRedirectAttr).length) )
+          {
+            this.scene.getField(true).filter(p => p !== this.pokemon).forEach(p => applyAbAttrs(RedirectMoveAbAttr, p, null, this.move.moveId, moveTarget));
+          }
```
[the remark this answers is not in the document]
Comment (f-raZ0R) [reply src/phases.ts:2218]: Does this look good?

File: src/data/move.ts [hunk 3]
```diff
@@ -2598,6 +2598,9 @@ const crashDamageFunc = (user: Pokemon, move: Move) => {
 
 export class TypelessAttr extends MoveAttr { }
 
+export class BypassRedirectAttr extends MoveAttr { }
+
+
```
Comment (bennybroseph) [inline src/data/move.ts:2603]: let's get some TSDoc documentation to explain why someone would what to use this attribute in the future

**Review by bennybroseph — CHANGES_REQUESTED** [review]

**Review by bennybroseph — APPROVED** [review]

**Review by bennybroseph — CHANGES_REQUESTED** [review]
When Propeller Tail/Stalwart activates, I'm not seeing the pop-up even though it is working. Are you seeing different?

**Comment by f-raZ0R** [comment]
This is intentional, as it does not make a pop-up in the official games either.

File: src/phases.ts [hunk 4]
```diff
@@ -2207,13 +2207,23 @@ export class MovePhase extends BattlePhase {
...
+          if ((this.pokemon.hasAbilityWithAttr(BlockRedirectAbAttr) || this.move.getMove().getAttrs(BypassRedirectAttr).length)) {
+              //If an ability prevented this move from being redirected, display its ability pop up.
+              if ((this.pokemon.hasAbilityWithAttr(BlockRedirectAbAttr) && !this.move.getMove().getAttrs(BypassRedirectAttr).length) && oldTarget != moveTarget.value) {
+                  this.scene.unshiftPhase(new ShowAbilityPhase(this.scene, this.pokemon.getBattlerIndex(), this.pokemon.getPassiveAbility().hasAttr(BlockRedirectAbAttr)));
+              }
+              moveTarget.value = oldTarget;
+          }
+          this.targets[0] = moveTarget.value;
+      }
+
```
Comment (bennybroseph) [inline src/phases.ts:2226]: looks like an extra return got added here

**Review by bennybroseph — CHANGES_REQUESTED** [review]

**Review by bennybroseph — APPROVED** [review]

**Comment by bennybroseph** [comment]
This is good to go in my opinion.

**Comment by bennybroseph** [comment]
This is good to go in my opinion.

**Review by Madmadness65 — APPROVED** [review]

**Pull request merged**